data/transactions/logic/assembler.go

```diff
@@ -2188,27 +2188,127 @@ func (dis *disassembleState) outputLabelIfNeeded() (err error) {
-	}
-	dis.nextpc = dis.pc + spec.Details.Size
 	out := spec.Name
-	for s := 1; s < spec.Details.Size; s++ {
-		b := uint(dis.program[dis.pc+s])
-		out += fmt.Sprintf(" %d", b)
+	pc := dis.pc + 1
+	for _, imm := range spec.Details.Immediates {
+		out += " "
+		switch imm.kind {
+		case immByte:
+			if pc >= len(dis.program) {
+				return "", fmt.Errorf("program end while reading immediate %s for %s",
+					imm.Name, spec.Name)
+			}
+			b := dis.program[pc]
+			if imm.Group != nil {
+				if int(b) >= len(imm.Group.Names) {
+					return "", fmt.Errorf("invalid immediate %s for %s: %d", imm.Name, spec.Name, b)
+				}
+				out += imm.Group.Names[b]
+			} else {
+				out += fmt.Sprintf("%d", b)
+			}
+			if spec.Name == "intc" && int(b) < len(dis.intc) {
+				out += fmt.Sprintf(" // %d", dis.intc[b])
+			}
+			if spec.Name == "bytec" && int(b) < len(dis.bytec) {
+				out += fmt.Sprintf(" // %s", guessByteFormat(dis.bytec[b]))
+			}
+
+			pc++
+		case immLabel:
+			offset := (uint(dis.program[pc]) << 8) | uint(dis.program[pc+1])
+			target := int(offset) + pc + 2
+			if target > 0xffff {
+				target -= 0x10000
+			}
+			var label string
+			if dis.numericTargets {
+				label = fmt.Sprintf("%d", target)
+			} else {
+				if known, ok := dis.pendingLabels[target]; ok {
+					label = known
+				} else {
+					dis.labelCount++
+					label = fmt.Sprintf("label%d", dis.labelCount)
+					dis.putLabel(label, target)
+				}
+			}
+			out += label
+			pc += 2
+		case immInt:
+			val, bytesUsed := binary.Uvarint(dis.program[pc:])
+			if bytesUsed <= 0 {
+				return "", fmt.Errorf("could not decode immediate %s for %s", imm.Name, spec.Name)
+			}
+			out += fmt.Sprintf("%d", val)
+			pc += bytesUsed
+		case immBytes:
+			length, bytesUsed := binary.Uvarint(dis.program[pc:])
+			if bytesUsed <= 0 {
+				return "", fmt.Errorf("could not decode immediate %s for %s", imm.Name, spec.Name)
+			}
+			pc += bytesUsed
+			end := uint64(pc) + length
+			if end > uint64(len(dis.program)) || end < uint64(pc) {
+				return "", fmt.Errorf("could not decode immediate %s for %s", imm.Name, spec.Name)
+			}
+			bytes := dis.program[pc:end]
+			out += fmt.Sprintf("0x%s // %s", hex.EncodeToString(bytes), guessByteFormat(bytes))
+			pc = int(end)
+		case immInts:
+			intc, nextpc, err := parseIntcblock(dis.program, pc)
+			if err != nil {
+				return "", err
+			}
+
+			dis.intc = append(dis.intc, intc...)
+			for i, iv := range intc {
+				if i != 0 {
+					out += " "
+				}
+				out += fmt.Sprintf("%d", iv)
+			}
+			pc = nextpc
+		case immBytess:
+			bytec, nextpc, err := parseBytecBlock(dis.program, pc)
+			if err != nil {
+				return "", err
+			}
+			dis.bytec = append(dis.bytec, bytec...)
+			for i, bv := range bytec {
+				if i != 0 {
+					out += " "
+				}
+				out += fmt.Sprintf("0x%s", hex.EncodeToString(bv))
+			}
+			pc = nextpc
+		default:
+			return "", fmt.Errorf("disDefault asked to disassemble complex immediate")
```

Comment:
maybe add `imm` and/or `imm.kind` to the error ?

Reply:
Agreed.